libeos-updater-util/flatpak.c

```diff
@@ -1294,11 +1301,37 @@ squash_ref_actions_ptr_array (GPtrArray *ref_actions  /* (element-type EuuFlatpa
 
   squashed_ref_actions = g_ptr_array_new_full (g_hash_table_size (hash_table),
                                                (GDestroyNotify) euu_flatpak_remote_ref_action_unref);
-  g_hash_table_iter_init (&hash_iter, hash_table);
 
-  while (g_hash_table_iter_next (&hash_iter, &key, &value))
-    g_ptr_array_add (squashed_ref_actions,
-                     euu_flatpak_remote_ref_action_ref (value));
+  /* Now iterate over the original array so we maintain its order while adding
+   * to the squashed array
+   */
+  for (i = 0; i < ref_actions->len; ++i)
+    {
+      EuuFlatpakRemoteRefAction *action = g_ptr_array_index (ref_actions, i);
+      EuuFlatpakRemoteRefAction *squashed_action_for_ref = NULL;
+
+      squashed_action_for_ref = g_hash_table_lookup (hash_table, action->ref->ref);
+      g_assert (squashed_action_for_ref != NULL);
+
+      /* Check that the action type matches so that
+       * [ install A, install B, uninstall A ] gets squashed into
+       * [ install B, uninstall A ] not [ uninstall A, install B ]
+       */
+      if (squashed_action_for_ref->type != action->type)
```

Comment:
I’m not sure about this either: I think it would fail in the situation where `ref_actions = [ install A, uninstall A, install B, install A, uninstall A ]`, though I haven’t worked the scenario through in detail.

Could we use a pointer comparison `if (squashed_action_for_ref != action) continue` here instead? Basically what we want to achieve is to ignore all the instances of any action for a given ref, unless it’s the `squashed_action_for_ref` for that ref.

Reply:
Agreed, pointer comparison is better. I also updated the unit test to cover this case.